File: 2020/python/day3.py
#!/bin/python3
input = open("../input/day3").readlines()

width = len(input[0]) - 1  # readLines adds whitespace at the end??
xpos = 0
ctr = 0

for line in input:
    if line[xpos] == '#':
        ctr = ctr + 1
    xpos = (xpos + 3) % width

print(ctr)

File: 2020/kotlin/day3.kt
import java.io.File

private fun part2(input: List<String>): Int {
    val width = input[0].length
    val count = { x: Int, y: Int ->
        var pos = Pair(0, 0)
        var ctr = 0
        while (pos.second < input.size) {
            if (input[pos.second][pos.first] == '#') ctr++
            pos = Pair((pos.first + x) % width, (pos.second + y))
        }
        ctr
    }
    return count(1, 1) * count(3, 1) * count(5, 1) * count(7, 1) * count(1, 2)
}

private fun part1(input: List<String>): Int {
    var xpos = 0
    val width = input[0].length
    return input.count {
        (it[xpos] == '#').also { xpos = (xpos + 3) % width }
    }
}

fun day3() {
    val input = File("2020/input/day3").readLines()

    println(part1(input))
    println(part2(input))
}